### DISCUSSION TITLE
Congratulations and a little question!

### DISCUSSION THREAD

**Discussion by frederikhors** [dpc/sniper#3] [Q&A]
First of all: **thank you very much for everything you wrote**.
Unlike what you wrote in one of your articles, there are people here who have greatly appreciated your work, your ideas and your blog and we hope you never stop writing new posts.

I have been using Rust for a few days and - coming from Golang - I have tried to "literally translate" the code getting stuck in the nested structs and infinite references cage (for which I have opened both [this](https://stackoverflow.com/questions/74127312/how-can-i-use-nested-structs) and [this](https://users.rust-lang.org/t/how-can-i-use-nested-structs/82971)) also trying to use SeaORM (to which I asked [this](https://github.com/SeaQL/sea-orm/discussions/1139)).

I started hacking with "`Box::new()`" and after a bit of experimenting I also succeeded. BUT I DON'T LIKE THE RESULT AT ALL!

Then I discovered you and data-oriented design.

As I'm in a hurry to finish this step in Rust I have not studied all the links you have recommended on your site (I will do so shortly) but I have tried to thoroughly study the use of `BTreeMap` that you recommend.

Also reading the (beautiful) [sniper code](https://github.com/dpc/sniper) I have a doubt: how do you keep the `BTreeMap` clean over time? I see no `.clear()` nor `.remove()` calls.

My idea is to use the `BTreeMap` as a "momentary mirror" of the DB data that I need for the operations I have to do right now. **Did I get it wrong or right?**

But after I'm done with those operations do I have to delete the data in the `BTreeMap`?

Assuming dozens of operations per second, all on different DB records, I can imagine that the `BTreeMap` can almost immediately take on a large size. I'm doing it wrong? What do you suggest?

Thanks again for everything and please don't stop!

**Comment by dpc** [comment]
> First of all: **thank you very much for everything you wrote**.

Thanks. I'm happy that it is useful!

> I have a doubt: how do you keep the BTreeMap clean over time? I see no .clear() nor .remove() calls.

Can you specify which `BTreeMap` do you mean? `InMemoryProgressTracker` is keyed on service id, so there shouldn't be all that many of them (services) in one program. `InMemoryBiddingStateStore` is per hypothetical item so it could grow with time.

Generally both are just "stores" and `InMemory...` is a in-memory implementation. They could be `Postgres...` and store everything in a database and be able to be much larger.

**Reply by frederikhors** [reply]
_I'm even more confused after your answer._

So far I have used a clean/hexagonal approach like this:

1. port - (HTTP (Rest API)) -> Incoming request - UPDATE /player calls:

1. service -> /handle_request_update_player calls:

1. adapter - (Postgres DB) -> query existing player:

    the adapter `mod` has structs like:

    ```rust
    // players table on DB
    type Player struct {
      id string
      name string
      Team Option<Box<Team>>
      //...
    }

    // teams table on DB
    type Team struct {
      id string
      Coach Option<Box<Coach>>
      //...
    }
    ```

    which are filled with (tabular) data from Postgres DB

1. these structs from the `adapter` are converted to domain structs which are roughly the same adapter structs (except the fields that are not needed or are different)

1. now I apply business logic/validation on domain structs rejecting if any error or proceeding if everything is ok

1. the domain structs are then converted **AGAIN** to DB PG adapter structs

1. the adapter persists those structs in the DB and return converted domain structs **AGAIN**

1. the service converts **AGAIN** the domain structs to port/interface HTTP structs (optional step in case of structs 1 to 1 equal)

1. the port/interface return the final value

All this is really cumbersome and wasteful. Not considering the serious problem of having `Box` for all the nested structs.

So - reading your blog - I thought of adding an intermediate layer (which for now would be the `BTreeMap`) - shared with all the levels - that can contain the dry data (coming from the DB adapter) to work on and which are then transformed where needed. In order to avoid all those heavy conversions of the current approach.

**Is this what you are suggesting with your articles (among other things)?**

If the answer is **yes** now you can better understand what I mean when I talk about **cleaning the `BTreeMap`**.

If the answer is **no** then I did not understand what you mean and I would like it if you explain it differently and thank you very much in advance.

**Reply by frederikhors** [reply]
@dpc Don't abandon me... please... :smile:

**Reply by dpc** [reply]
I'm afraid I'm actually not an expert, but just to share thoughts:

I don't understand why you have Boxes around your data? To allocate on the heap? In sniper I was mostly focused on figuring out the way to abstract away implementation of: persistance (database threadpool), connection, and transaction.

> 7. the adapter persists those structs in the DB and return converted domain structs AGAIN

Not sure why conversion back to something that was already available is needed.

As for conversions - if you have different representations of each entity in different layers, then conversions might be necessary, thought when possible, might be easier to avoid. These conversions should not be "heavy" though. Most fields should not need any conversion at all (string field turns into string field) so they will be mostly a no-op at runtime.

I don't understand what do you need that `BTreeMap` :) . In sniper the `BTreeMap` was used mostly as a in-memory implementation of stores (/database) for things like unit-tests. Seems like you're trying to build your own cache or something?

**Reply by frederikhors** [reply]
> I don't understand why you have Boxes around your data?

This is why: https://www.sea-ql.org/blog/2022-05-14-faq-01/#q01-why-seaorm-does-not-nest-objects-for-parent-child-relation.

> Seems like you're trying to build your own cache or something?

I'm using a struct with `HashMap` fields, one for each related entity and I'm passing it where needed instead of `structs` with `Box`es.

**Is this what you recommend in your articles?**
